L1Trigger/TrackFindingTMTT/src/L1KalmanComb.cc

```diff
@@ -0,0 +1,1828 @@
+
+    std::vector<const StubCluster *> stubcls;
+
+    for (unsigned j_layer = 0; j_layer < 16; j_layer++) {
+      std::vector<const Stub *> layer_stubs;
+      for (unsigned i = 0; i < stubs.size(); i++) {
+        const Stub *stub = stubs.at(i);
+        if (stub->layerId() == LayerId[j_layer]) {
+          layer_stubs.push_back(stub);
+        }
+      }
+
+#ifdef MERGE_STUBS
+      if (LayerId[j_layer] < 10)
+        sort(layer_stubs.begin(), layer_stubs.end(), orderStubsByZ);  // barrel
+      else
+        sort(layer_stubs.begin(), layer_stubs.end(), orderStubsByR);  // endcap
+#endif
+
+      for (unsigned i = 0; i < layer_stubs.size(); i++) {  // Stubs in single layer, ordered by z or r.
+
+        std::vector<const Stub *> stubs_for_cls;
+        stubs_for_cls.push_back(layer_stubs.at(i));
+
+#ifdef MERGE_STUBS
+        while (layer_stubs.at(i) != layer_stubs.back()) {
+          if (isOverlap(layer_stubs.at(i), layer_stubs.at(i + 1), TYPE_NORMAL)) {
+            stubs_for_cls.push_back(layer_stubs.at(i + 1));
+            if (getSettings()->kalmanFillInternalHists())
+              hnmergeStub_->Fill(0);
+            i++;
+          } else
+            break;
+        }
+#endif
+
+        if (getSettings()->kalmanFillInternalHists()) {
+          if (tpa && tpa->useForAlgEff()) {
+            if (stubs_for_cls.size() > 1) {
+              std::set<const TP *> s_tps = stubs_for_cls.at(0)->assocTPs();
+              if (s_tps.find(tpa) != s_tps.end()) {
+                const Stub *sa = stubs_for_cls.front();
+                const Stub *sb = stubs_for_cls.back();
+
+                double drphi = fabs(sa->r() * reco::deltaPhi(sa->phi(), sectorPhi()) -
+                                    sb->r() * reco::deltaPhi(sb->phi(), sectorPhi()));
+                double dz = fabs(sa->z() - sb->z());
+                double dr = fabs(sa->r() - sb->r());
+                TString hname;
+                if (LayerId[j_layer] < 10) {
+                  hname = Form("hBarrelStubMaxDistanceLayer%02d", LayerId[j_layer]);
+
+                  if (hBarrelStubMaxDistanceMap.find(hname) == hBarrelStubMaxDistanceMap.end()) {
+                    cout << hname << " does not exist." << endl;
+                  } else {
+                    hBarrelStubMaxDistanceMap[hname]->Fill(drphi, dz);
+                  }
+                } else {
+                  hname = Form("hEndcapStubMaxDistanceRing%02d", sa->endcapRing());
+
+                  if (hEndcapStubMaxDistanceMap.find(hname) == hEndcapStubMaxDistanceMap.end()) {
+                    cout << hname << " does not exist." << endl;
+                  } else {
+                    hEndcapStubMaxDistanceMap[hname]->Fill(drphi, dr);
+                  }
+                }
+              }
+            }
+          }
+        }
+
+        // dl error now disabled
+        StubCluster *stbcl = new StubCluster(stubs_for_cls, sectorPhi(), 0);
+        stbcl_list_.push_back(stbcl);
+        stubcls.push_back(stbcl);
```

Comment:
Please consider using `unique_ptr` for the one who owns it.

Reply:
After simplifying the KF, the StubCluster class has been deleted.